### PR TITLE
add pair generator

### PR DESCRIPTION
**Pull request by faneshion**
#266

### CODE REVIEW

File: matchzoo/generators/pair_generator.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        >>> assert x is not None
+        >>> assert y is not None
+
+    """
+
+    def __init__(
+        self,
+        inputs: datapack.DataPack,
+        num_neg: int=1,
+        dup_time: int=4,
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: maybe it's better to change the variable name, such as `num_dup`, use the same naming convention as `num_neg`

File: matchzoo/generators/pair_generator.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        :return: A batch of transformed samples.
+        """
+        batch_x = {}
+        batch_y = []
+        for pos_sample, neg_samples in self.pairs[index_array]:
+            batch_y.append(pos_sample['label'])
+            for _, neg_sample in neg_samples[0].iterrows():
+                batch_y.append(neg_sample['label'])
+        num_ids = len(index_array) * (self._num_neg + 1)
+        batch_x['ids'] = [[] for i in range(num_ids)]
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: `[[]] * len(num_ids)`

File: matchzoo/generators/pair_generator.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        for key in self._columns:
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: `for column in self._columns` is better and more readable.

File: matchzoo/generators/pair_generator.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            idx = 0
+            for pos_sample, neg_samples in self.pairs[index_array]:
+                batch_x[key].append(self.content[pos_sample[key]])
+                batch_x['ids'][idx].append(pos_sample[key])
+                idx += 1
+                for _, neg_sample in neg_samples[0].iterrows():
+                    batch_x[key].append(self.content[neg_sample[key]])
+                    batch_x['ids'][idx].append(neg_sample[key])
+                    idx += 1
+            batch_x[key] = np.array(batch_x[key])
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: line 106-line 126 need refactoring.

File: matchzoo/generators/pair_generator.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                pos_samples = pd.concat([pos_samples] * self._dup_time)
+                neg_samples = group[group.label < label]
+                for _, pos_sample in pos_samples.iterrows():
+                    pos_sample = pd.DataFrame([pos_sample])
+                    if len(neg_samples) >= self._num_neg:
+                        neg_sample = neg_samples.sample(self._num_neg,
+                                                        replace=False)
+                        pairs.extend((pos_sample, neg_sample))
+                        num_pairs += 1
+        return num_pairs, pd.concat(pairs, ignore_index=True)
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: why `num_pairs` and add 1 per iteration? why not `return len(pairs)`?

File: matchzoo/generators/pair_generator.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        ...                         columns=['id_left', 'id_right', 'label'])
+        >>> left = pd.DataFrame(left, columns=['id_left', 'text_left'])
+        >>> left.set_index('id_left', inplace=True)
+        >>> right = pd.DataFrame(right, columns=['id_right', 'text_right'])
+        >>> right.set_index('id_right', inplace=True)
+        >>> input = datapack.DataPack(relation=relation,
+        ...                           left=left,
+        ...                           right=right
+        ... )
+        >>> from matchzoo.generators import PairGenerator
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: do not import because you're here.

File: matchzoo/generators/pair_generator.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        >>> generator = PairGenerator(input, 1, 1, 1, 'train', True)
+        >>> assert generator
+        >>> len(generator)
+        2
+        >>> x, y = generator[0]
+        >>> assert x is not None
+        >>> assert y is not None
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: do not write `assert` in doctest, use:
```python
>>> x
[1,2,3]
```

File: tests/unit_test/generators/test_pair_generator.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                              num_neg=1,
+                              dup_time=2,
+                              batch_size=batch_size,
+                              stage='train',
+                              shuffle=shuffle)
+    assert len(generator) == 4
+    for idx, (x, y) in enumerate(generator):
+        if idx == len(generator):
+            break
+        assert x is not None
```
Comment (bwanglzu) [inline tests/unit_test/generators/test_pair_generator.py]: mentioned during the discussion, not actually testing .

File: tests/unit_test/generators/test_pair_generator.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        assert y is not None
+
+def test_pair_generator_multi(x):
+    """Test pair generator with multiple negative sample."""
+    shuffle = False
+    batch_size = 1
+    generator = PairGenerator(x, 2, 2, batch_size, 'test', shuffle)
+    assert len(generator) == 2
+    for x, y in generator:
+        assert x is not None
```
Comment (bwanglzu) [inline tests/unit_test/generators/test_pair_generator.py]: same

File: matchzoo/generators/pair_generator.py (hunk 5 above)
Comment (wsdm2019-dmpa) [reply matchzoo/generators/pair_generator.py]: Here, len(pairs) not equal to num_pairs. For case `num_neg = 2`, then, `pairs = [pos0, neg00, neg01, pos1, neg10, neg11, ...]`. Thus, `num_pairs = len(pairs) / (num_neg + 1)`.

File: matchzoo/generators/pair_generator.py (hunk 6 above)
Comment (wsdm2019-dmpa) [reply matchzoo/generators/pair_generator.py]: ok

File: matchzoo/generators/pair_generator.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    Pair generator can be used only for ranking.
+
+    Examples:
+        >>> np.random.seed(111)
+        >>> relation = [['qid0', 'did0', 0],
+        ...             ['qid0', 'did1', 1],
+        ...             ['qid0', 'did2', 2]
+        ... ]
+        >>> columns = ['id_left', 'id_right', 'label']
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: defined but never used.

File: matchzoo/generators/pair_generator.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        >>> left = pd.DataFrame(left, columns=['id_left', 'text_left'])
+        >>> left.set_index('id_left', inplace=True)
+        >>> right = pd.DataFrame(right, columns=['id_right', 'text_right'])
+        >>> right.set_index('id_right', inplace=True)
+        >>> input = datapack.DataPack(relation=relation,
+        ...                           left=left,
+        ...                           right=right
+        ... )
+        >>> generator = PairGenerator(input, 1, 1, 1, 'train', True)
+        >>> assert generator
```
Comment (bwanglzu) [inline matchzoo/generators/pair_generator.py]: remove

File: matchzoo/generators/point_generator.py [hunk 12]
```diff
@@ -32,8 +32,10 @@ class PointGenerator(engine.BaseGenerator):
         >>> task = tasks.Classification()
         >>> generator = PointGenerator(input, task, 1, 'train', False)
         >>> x, y = generator[0]
-        >>> x['text_left'].tolist()
-        [[1, 2]]
+        >>> assert x['text_left'].tolist() == [[1, 2]]
+        >>> assert x['text_right'].tolist() == [[2, 3]]
+        >>> assert x['ids'].tolist() == [['qid0', 'did0']]
+        >>> assert y.tolist() == [[0., 1.]]
```
Comment (bwanglzu) [inline matchzoo/generators/point_generator.py]: do not assert in doctest

File: matchzoo/generators/pair_generator.py (hunk 11 above)
Comment (wsdm2019-dmpa) [reply matchzoo/generators/pair_generator.py]: done~

File: matchzoo/generators/pair_generator.py (hunk 10 above)
Comment (wsdm2019-dmpa) [reply matchzoo/generators/pair_generator.py]: removed~

File: matchzoo/generators/point_generator.py (hunk 12 above)
Comment (wsdm2019-dmpa) [reply matchzoo/generators/point_generator.py]: ok~

**Pull request merged**